File: 300.600/404.SumOfLeftLeaves.cpp
#include <iostream>
using namespace std;

struct TreeNode {
    int val{};
    TreeNode* left{};
    TreeNode* right{};
    TreeNode() {}
    TreeNode(int x) : val(x) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class SumOfLeftLeaves {
public:
    int sumOfLeftLeaves(TreeNode* root) {
        ret = 0;
        if (root)
            dfs(root, false);
        return ret;
    }
private:
    void dfs(TreeNode* root, bool left) {
        if (root->left == nullptr && root->right == nullptr) {
            if (left)
                ret += root->val;
            return;
        }
        if (root->left)
            dfs(root->left, true);
        if (root->right)
            dfs(root->right, false);
    }
private:
    int ret{};
};

void drop(TreeNode* root) {
    if (!root)
        return;
    drop(root->left);
    drop(root->right);
    delete root;
}

int main(int argc, char const* argv[]) {
    SumOfLeftLeaves cls;
    TreeNode* head = new TreeNode(3);
    head->left = new TreeNode(9);
    head->right = new TreeNode(20);
    head->right->left = new TreeNode(15);
    head->right->right = new TreeNode(7);
    cout << cls.sumOfLeftLeaves(head) << " -> 24" << endl;
    drop(head);
    return 0;
}

File: go/300.500/404.SumOfLeftLeaves.go
package main

import "fmt"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func sumOfLeftLeaves(root *TreeNode) int {
	ret := 0
	var dfs func(root *TreeNode, left bool)
	dfs = func(root *TreeNode, left bool) {
		if root.Left == nil && root.Right == nil {
			if left {
				ret += root.Val
			}
			return
		}
		if root.Left != nil {
			dfs(root.Left, true)
		}
		if root.Right != nil {
			dfs(root.Right, false)
		}
	}
	if root != nil {
		dfs(root, false)
	}
	return ret
}

func main() {
	head := &TreeNode{Val: 3}
	head.Left = &TreeNode{Val: 9}
	head.Right = &TreeNode{Val: 20}
	head.Right.Left = &TreeNode{Val: 15}
	head.Right.Right = &TreeNode{Val: 7}
	fmt.Println(sumOfLeftLeaves(head), 24)
}